File: q0070/q0070_recursive.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = [-1] * (n+1)
        return self.dpHelper(n, memo)
    
    def dpHelper(self, n, memo): 
        if n == 0: return 1
        if n == 1: return 1
        if memo[n] != -1: return memo[n]
        res = self.dpHelper(n-1, memo) + self.dpHelper(n-2, memo)
        memo[n] = res
        return res

File: q0070/q0070_recursive.java
class Solution {
    public int climbStairs(int n) {
        int[] memo = new int[n + 1];
        Arrays.fill(memo, -1);
        return dpHelper(n, memo);
    }

    // Let dpHelper return the number of ways to reach n
    private int dpHelper(int n, int[] memo) {

        // Base cases
        if (n == 0)
            return 1;
        if (n == 1)
            return 1;

        // Memoized case
        if (memo[n] != -1)
            return memo[n];

        // Recursive case
        int res = dpHelper(n - 1, memo) + dpHelper(n - 2, memo);

        // Memoization
        memo[n] = res;
        return res;
    }
}